lesson/outline/learning-standards.md

```diff
@@ -0,0 +1,256 @@
+
+In this example, the `userEvent.type()` method is used to simulate a user typing the text `'Hello World!'` into the rendrered text box.
+
+```js
+import React from 'react'
+import {render, screen} from '@testing-library/react'
+import userEvent from '@testing-library/user-event'
+import '@testing-library/jest-dom';
+
+it('should change the text of the input', () => {
+  render(<textarea />);
+
+  userEvent.type(screen.getByRole('textbox'), 'Hello World!')
+  expect(screen.getByRole('textbox')).toHaveValue('Hello World!')
+})
+```
+
+### Tags
+
+JavaScript, React, Unit Testing
+
+### Additional Notes / Resources Used
+
+- https://github.com/testing-library/user-event#typeelement-text-options
+
+<hr>
+
+## queryBy* variants
+
+### Learning Standard Text
+
+When using the React Testing Library (RTL) to determine if an element is NOT present in the rendered DOM, the `screen.queryBy*` variants should be used over their `screen.getBy*` counterparts.
+
+If the queried element produces a null set, the `screen.getBy*` variants will throw an error before the assertion can be made whereas the `screen.queryBy*` variants simply return `null`.
+
+In this example, using `screen.getByRole('foobar')` will result in an error and a failing test while using `screen.queryByRole('foobar')` will result in a passing test.
+
+```js
+// Assume that there is no element with the role 'foobar'
+
+// getByRole('foobar') will throw an error if no element is found
+expect(screen.getByRole('foobar')).toBeNull();
+
+// queryByRole simply returns null if no element is found
+expect(screen.queryByRole('foobar')).toBeNull();
+```
+
+### Tags
+
+JavaScript, React, Unit Testing
+
+### Additional Notes / Resources Used
+
+- https://testing-library.com/docs/react-testing-library/example-intro
+- https://www.robinwieruch.de/react-testing-library
+- https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#using-query-variants-for-anything-except-checking-for-non-existence
+
+<hr>
+
+## findBy* variants
+
+### Learning Standard Text
+
+When using the React Testing Library (RTL) to query the rendered DOM for an element that will appear as a result of an asynchronous action (for example, after some API fetch occurs), the `screen.findBy*` variants should be used instead of the `waitFor()` method and the `screen.getBy*` variants.
+
+In this example, a header with the test `'Goodbye'` will appear after a Promise resolves. While both approaches will accurately find this element, the approach that uses `screen.findByText()` is more concise and can produce better error messages in the event that the element is not found.
+
+```js
+// Wait for an element with the test 'Goodbye' to appear before proceeding.
+const header = await waitFor(() =>
+  screen.getByText('Goodbye')
+)
+
+// This does the same as above, but more succinctly and with better error logging
+const header = await screen.findByText('Goodbye');
+```
```

Comment:
```suggestion
```js
it ('should find an async element', async () => { 
  // Wait for an element with the test 'Goodbye' to appear before proceeding.
  const header = await waitFor(() =>
    screen.getByText('Goodbye')
  )

  // This does the same as above, but more succinctly and with better error logging
  const header = await screen.findByText('Goodbye');
  
  expect(header).toBeInTheDocument();
});
```
```